File: Backend/app/services/product_service.py
from app.repositories.product_repository import ProductRepository
from app.schemas.product import ProductCreate, ProductResponse
from typing import List
from uuid import UUID


class ProductService:
    def __init__(self, repository: ProductRepository):
        self.repository = repository

    async def create_product(self, product_data: ProductCreate) -> ProductResponse:
        product = await self.repository.create(product_data.model_dump())
        return ProductResponse.model_validate(product)

    async def get_product(self, product_id: UUID) -> ProductResponse:
        product = await self.repository.get_by_id(product_id)
        if not product:
            raise ValueError("Product not found")
        return ProductResponse.model_validate(product)

    async def get_all_products(self) -> List[ProductResponse]:
        products = await self.repository.get_all()
        return [ProductResponse.model_validate(p) for p in products]

File: Frontend/src/services/product.service.ts
import { apiClient } from '../api/client';

export interface Product {
  id: string;
  name: string;
  unit_price: number;
}

export interface ProductCreate {
  name: string;
  unit_price: number;
}

export class ProductService {
  private readonly BASE_PATH = '/products';

  async getAllProducts(): Promise<Product[]> {
    const response = await apiClient.get<Product[]>(this.BASE_PATH);
    return response.data;
  }

  async getProduct(productId: string): Promise<Product> {
    const response = await apiClient.get<Product>(`${this.BASE_PATH}/${productId}`);
    return response.data;
  }

  async createProduct(product: ProductCreate): Promise<Product> {
    const response = await apiClient.post<Product>(this.BASE_PATH, product);
    return response.data;
  }
}

export const productService = new ProductService();